execution/gethexec/espresso_finality_node.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/offchainlabs/nitro/arbos"
+	"time"
+
+	"github.com/offchainlabs/nitro/arbos/arbostypes"
+	"github.com/offchainlabs/nitro/arbos/l1pricing"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+)
+
+var (
+	retryTime = time.Second * 1
+)
+
+/*
+Espresso Finality Node creates blocks with finalized hotshot transactions
+*/
+type EspressoFinalityNode struct {
+	stopwaiter.StopWaiter
+
+	config     SequencerConfigFetcher
+	execEngine *ExecutionEngine
+	namespace  uint64
+
+	espressoClient  *espressoClient.Client
+	nextSeqBlockNum uint64
+}
+
+func NewEspressoFinalityNode(execEngine *ExecutionEngine, configFetcher SequencerConfigFetcher) *EspressoFinalityNode {
+	config := configFetcher()
+	if err := config.Validate(); err != nil {
+		panic(err)
+	}
+	return &EspressoFinalityNode{
+		execEngine:      execEngine,
+		config:          configFetcher,
+		namespace:       config.EspressoFinalityNodeConfig.Namespace,
+		espressoClient:  espressoClient.NewClient(config.EspressoFinalityNodeConfig.HotShotUrl),
+		nextSeqBlockNum: config.EspressoFinalityNodeConfig.StartBlock,
+	}
+}
+
+func (n *EspressoFinalityNode) createBlock(ctx context.Context) (returnValue bool) {
+	if n.nextSeqBlockNum == 0 {
+		latestBlock, err := n.espressoClient.FetchLatestBlockHeight(ctx)
+		if err != nil && latestBlock == 0 {
+			log.Warn("unable to fetch latest hotshot block", "err", err)
+			return false
+		}
+		log.Info("Started espresso finality node at the latest hotshot block", "block number", latestBlock)
+		n.nextSeqBlockNum = latestBlock
+	}
+
+	nextSeqBlockNum := n.nextSeqBlockNum
+	header, err := n.espressoClient.FetchHeaderByHeight(ctx, nextSeqBlockNum)
+	if err != nil {
+		arbos.LogFailedToFetchHeader(nextSeqBlockNum)
+		return false
+	}
+
+	arbTxns, err := n.espressoClient.FetchTransactionsInBlock(ctx, header.Height, n.namespace)
+	if err != nil {
+		arbos.LogFailedToFetchTransactions(header.Height, err)
+		return false
+	}
+	arbHeader := &arbostypes.L1IncomingMessageHeader{
+		Kind:        arbostypes.L1MessageType_L2Message,
+		Poster:      l1pricing.BatchPosterAddress,
+		BlockNumber: header.L1Head,
+		Timestamp:   header.Timestamp,
+		RequestId:   nil,
+		L1BaseFee:   nil,
+	}
+
+	// Deserialize the transactions and remove the signature from the transactions.
+	// Ignore the malformed transactions
+	txes := types.Transactions{}
+	for _, tx := range arbTxns.Transactions {
+		var out types.Transaction
+		// signature from teh data poster  is the first 65 bytes of a transaction
+		tx = tx[65:]
+		if err := out.UnmarshalBinary(tx); err != nil {
+			log.Warn("malformed tx found")
+			continue
+		}
+		txes = append(txes, &out)
+	}
+
+	hooks := arbos.NoopSequencingHooks()
+	_, err = n.execEngine.SequenceTransactions(arbHeader, txes, hooks, false)
+	if err != nil {
+		log.Error("espresso finality node: failed to sequence transactions", "err", err)
+		return false
+	}
+	n.nextSeqBlockNum += 1
```

Comment:
I think this line should be moved to `Start` function. That will be clearer I think